### PR TITLE
refactor(bundler): refactor extraction regexes

### PR DESCRIPTION
**Pull request by Djiit**
## Changes:

Break the line regex into mutliple one. Instead of capturing all the line at once, we ensure it's a `gem` line, then apply other regexes sequentially. A few benefits:

- Readability
- Support randomly ordered arguments (e.g. `gem 'foo', source: baz, require: 'bar'` and `gem 'foo', require: 'bar', source: baz` are now parsed the same way

Current tests are still passing.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/bundler/__snapshots__/extract.spec.ts.snap [hunk 1]
```diff
@@ -542,12 +542,15 @@ exports[`modules/manager/bundler/extract extractPackageFile() parse mastodon Gem
       },
     },
     {
+      "currentDigest": "54b17ba8c7d8d20a16dfc65d1775241833219cf2",
       "currentValue": "'~> 0.6'",
-      "datasource": "rubygems",
```
Comment (Djiit) [inline lib/modules/manager/bundler/__snapshots__/extract.spec.ts.snap:546]: Why this change? We were missing this one because of an un-handled line order. This refactor allow finer extraction, and allow for such fixes.

File: lib/modules/manager/bundler/extract.spec.ts [hunk 2]
```diff
@@ -171,14 +171,21 @@ describe('modules/manager/bundler/extract', () => {
   it('parses inline source in Gemfile', async () => {
     const sourceInlineGemfile = codeBlock`
       baz = 'https://gems.baz.com'
+      gem 'inline_gem'
       gem "inline_source_gem", source: 'https://gems.foo.com'
       gem 'inline_source_gem_with_version', "~> 1", source: 'https://gems.bar.com'
       gem 'inline_source_gem_with_variable_source', source: baz
+      gem "inline_source_gem_with_require_after", source: 'https://gems.foo.com', require: %w[inline_source_gem]
```
Comment (Djiit) [inline lib/modules/manager/bundler/extract.spec.ts:178]: Added a few combination to ensure futures changes won't break any known ordering.

File: lib/modules/manager/bundler/__snapshots__/extract.spec.ts.snap [hunk 3]
```diff
@@ -1606,7 +1609,7 @@ exports[`modules/manager/bundler/extract extractPackageFile() parses rails Gemfi
       },
     },
     {
-      "currentValue": "">= 3.0.5", "< 3.2"",
+      "currentValue": "">= 3.0.5"",
```
Comment (rarkins) [inline lib/modules/manager/bundler/__snapshots__/extract.spec.ts.snap:1612]: This seems like a regression?

Comment (Djiit) [reply lib/modules/manager/bundler/__snapshots__/extract.spec.ts.snap]: Good catch. I updated the regex and added a dedicated test.

**Review by rarkins — APPROVED** [review]

**Review by secustor — APPROVED** [review]
